### PR TITLE
Rework SysvarCache implementation

### PR DESCRIPTION
**Pull request by 2501babe**
#### Problem

we have proposed a syscall, `sol_get_sysvar`/`SyscallGetSysvar`, a generic mechanism to fetch arbitrary slices of account data for a specially selected list of sysvars. this is intended primarily to enable the bpf versions of the stake and vote programs to access individual parts of the stake history and slot hashes sysvars, but it also functions as a drop-in replacement for all existing non-deprecated sysvar syscalls

see discussion in [SIMD-0127](https://github.com/solana-foundation/solana-improvement-documents/pull/127) for (much) more context

#### Summary of Changes

we change `SysvarCache` contents from arc-wrapped objects to u8 vectors containing the exact onchain account data for the sysvars. small (non-vector) sysvars are now deserialized on-demand as the cost is trivial and should not be incurred more than once per instruction. large (stake history and slot hashes) sysvars are kept around in the old format for the sake of the stake and vote programs, but will be removed after native to bpf conversion. we keep the deprecated fees and recent blockhashes sysvars, these can likely also be removed after native to bpf conversion

we wrap our return values in `Arc` to minimize change surface and preserve the existing interface

this pr is part one of two: a second pr will implement the syscall. a third and fourth will actually use the syscall for getting slices of `StakeHistory` and `SlotHashes`

### CODE REVIEW

**Comment by 2501babe** [comment]
@buffalojoec ive added you as a reviewer while this is still in draft, please let me know if you think there should be any final changes. im not planning of modifying this code (except in response to reviews) any further, other than removing the stake history stub method. i need to write tests still tho is why im not moving out of draft

File: sdk/program/src/syscalls/definitions.rs
Comment (buffalojoec) [inline sdk/program/src/syscalls/definitions.rs]: Cool. I like the organization here.

File: programs/bpf_loader/src/syscalls/sysvar.rs
Comment (buffalojoec) [inline programs/bpf_loader/src/syscalls/sysvar.rs]: We should make sure these parameters are in the exact same order as specified in the SIMD.
https://github.com/solana-foundation/solana-improvement-documents/pull/127/files#diff-c0cf44e3505bfa26cded9fe200049fc086f29f0f4f7bb6f2dc348cdbc6e96988R139-R144

File: programs/bpf_loader/src/syscalls/sysvar.rs
Comment (buffalojoec) [inline programs/bpf_loader/src/syscalls/sysvar.rs]: Just gotta bring this bad boy to parity with the SIMD specification for CUs.
https://github.com/solana-foundation/solana-improvement-documents/pull/127/files#diff-c0cf44e3505bfa26cded9fe200049fc086f29f0f4f7bb6f2dc348cdbc6e96988R173-R175

File: sdk/program/src/program_stubs.rs
Comment (buffalojoec) [inline sdk/program/src/program_stubs.rs]: I think the program stubs should match the parameter ordering on the syscall, too (maybe they _have_ to).

File: sdk/program/src/syscalls/definitions.rs
Comment (buffalojoec) [inline sdk/program/src/syscalls/definitions.rs]: Just one more spot to check the parameter ordering!

File: program-runtime/src/sysvar_cache.rs
Comment (buffalojoec) [inline program-runtime/src/sysvar_cache.rs]: I'm thinking mapping this error to `UnsupportedSysvar` is problematic, but I'm also aware that _in theory_ this account data should never be malformed.

However, the inability to deserialize from within the new syscall (as a result of malformed bytes) is not actually specified in the SIMD (for this reason). Do you have thoughts on whether or not we should use something else here?

File: program-runtime/src/sysvar_cache.rs
Comment (buffalojoec) [inline program-runtime/src/sysvar_cache.rs]: Just wanted to point out that as a result of https://github.com/anza-xyz/agave/pull/249, we can introduce zero-copy (non-bincode) getters, too. Maybe something like `get_clock_ref() -> &Clock`. Could be useful for parts of the runtime, and moving away from object getters who use bincode.

File: programs/bpf_loader/src/syscalls/sysvar.rs
Comment (buffalojoec) [inline programs/bpf_loader/src/syscalls/sysvar.rs]: Your sysvar cache implementation seems to handle the VM-abort control flow by throwing errors, but I think you're missing the two graceful results specified:
https://github.com/solana-foundation/solana-improvement-documents/pull/127/files#diff-c0cf44e3505bfa26cded9fe200049fc086f29f0f4f7bb6f2dc348cdbc6e96988R158-R164

File: programs/bpf_loader/src/syscalls/sysvar.rs
Comment (buffalojoec) [inline programs/bpf_loader/src/syscalls/sysvar.rs]: Ha, strange. Why do we have situations with non-zero alignment bytes?

File: program-runtime/src/sysvar_cache.rs
Comment (buffalojoec) [inline program-runtime/src/sysvar_cache.rs]: I might be missing something here, but since the docs say this test is supposed to assert the sysvar cache is keeping all of the bytes from the account data, should it be able to check & handle a case where, say, the account data is larger than `T::size_of()`?

File: programs/bpf_loader/src/syscalls/mod.rs
Comment (buffalojoec) [inline programs/bpf_loader/src/syscalls/mod.rs]: Heh, this test is great.

File: program-runtime/src/sysvar_cache.rs
Comment (buffalojoec) [inline program-runtime/src/sysvar_cache.rs]: In this function and and `get_sysvar_obj`, when `sysvar_id_to_buffer` returns `None`, you return the error `UnsupportedSysvar`. Why not just hoist this error into `sysvar_id_to_buffer`?

**Review by buffalojoec — COMMENTED** [review]
Nice, looking slick!!

I left a handful of comments, mostly questions and relating to the specifics of the SIMD.

We just want to make sure this new syscall is totally aligned with the spec. Now that the SIMD is approved, hopefully we have a stable spec now!

**Comment by buffalojoec** [comment]
> i am of the opinion that breaking the interfaces of the sysvar cache object getters to remove the `Arc` is fine, because nothing outside the monorepo has any reasonable use for the sysvar cache.

I think because you're deserializing on read, it's thread-safe without `Arc`.

File: programs/bpf_loader/src/syscalls/sysvar.rs
Comment (2501babe) [reply programs/bpf_loader/src/syscalls/sysvar.rs]: bincode doesnt zero them presumably for efficiency

```rust
use serde::{Serialize, Deserialize};

#[repr(C)]
#[derive(Clone, Copy, Serialize, Deserialize)]
struct Thing { 
    qword: u64,
    byte: u8,
}

fn main() {
    let thing1 = Thing { qword: 1, byte: 2 };
    let thing2 = bincode::deserialize::<Thing>(&bincode::serialize(&thing1).unwrap()).unwrap();
    let thing2_ptr = &thing2 as *const Thing as *const u8;

    for i in 0..std::mem::size_of::<Thing>() {
        print!("{} ", unsafe { *thing2_ptr.wrapping_add(i) });
    }
    println!("");
}
```

```
hana@laptop:bctest$ cargo run
    Finished dev [unoptimized + debuginfo] target(s) in 0.00s
     Running `target/debug/bctest`
1 0 0 0 0 0 0 0 2 150 148 54 0 0 0 0
```

File: programs/bpf_loader/src/syscalls/sysvar.rs
Comment (2501babe) [reply programs/bpf_loader/src/syscalls/sysvar.rs]: ty, impled this before there was a defined interface

File: programs/bpf_loader/src/syscalls/sysvar.rs
Comment (2501babe) [reply programs/bpf_loader/src/syscalls/sysvar.rs]: just curious what is `(32 / cpi_per_u)` meant to charge for?

Comment (buffalojoec) [reply programs/bpf_loader/src/syscalls/sysvar.rs]: It's a pattern in a lot of the other syscalls for byte-wise charging, and in this case it's for translating the 32 bytes of the sysvar ID.

**Comment by 2501babe** [comment]
>I think because you're deserializing on read, it's thread-safe without Arc.

yea every caller gets its own copy of each sysvar now, do `Arc` doesnt serve a purpose in that regard. i was just thinking about whether we ought to keep the return type the same. i dont believe anything outside of this repo should have any reason to call these functions, but i wonder if its unwise to make that assumption

File: program-runtime/src/sysvar_cache.rs
Comment (2501babe) [reply program-runtime/src/sysvar_cache.rs]: hmm, it just seemed cleaner to do it this way. since we want to take a reference to the `Vec<u8>` and only copy what we need from it, if `sysvar_id_to_buffer` returned a `Result<&Vec<u8>, _>`, we would have to unwrap the `Option` in `sysvar_id_to_buffer` anyway

File: program-runtime/src/sysvar_cache.rs
Comment (2501babe) [reply program-runtime/src/sysvar_cache.rs]: thats an interesting point i hadnt considered. i guess we can also cover that case by making `size` bigger than it needs to be. the specific case i was trying to cover was just that if you roundtrip a `StakeHistory` or `SlotHashes` that has less than `MAX_ENTRIES` through bincode it truncates the `Vec`

File: program-runtime/src/sysvar_cache.rs
Comment (2501babe) [reply program-runtime/src/sysvar_cache.rs]: `get_sysvar_obj` is unrelated to the new syscall and merely preserves behavior for the existing sysvar syscalls. in the old version of the cache, if the data passed into `fill_missing_entries` fails to parse, then the cache entry is silently left as `None`, so when the getter is called `UnsupportedSysvar` will be returned. we also check that the data deserializes before keeping the vector, to preserve this behavior

File: program-runtime/src/sysvar_cache.rs
Comment (2501babe) [reply program-runtime/src/sysvar_cache.rs]: oh thats an interesting thought. im not sure if i want to add complexity by incorporating that into this pr tho

File: programs/bpf_loader/src/syscalls/sysvar.rs
Comment (2501babe) [reply programs/bpf_loader/src/syscalls/sysvar.rs]: will get to this in a separate commit tonight since ill wanna write tests too

File: program-runtime/src/sysvar_cache.rs
Comment (2501babe) [reply program-runtime/src/sysvar_cache.rs]: incidentally i realize this means we do in fact want to use vectors rather than arrays in the cache (remember discussing this in the past) because we have no access to the actual size of the onchain data in code

File: program-runtime/src/sysvar_cache.rs
Comment (buffalojoec) [reply program-runtime/src/sysvar_cache.rs]: Agreed it probably doesn't belong in this PR, but worth considering in the future!

File: program-runtime/src/sysvar_cache.rs
Comment (buffalojoec) [reply program-runtime/src/sysvar_cache.rs]: Makes sense to me, sounds good!

File: program-runtime/src/sysvar_cache.rs
Comment (buffalojoec) [reply program-runtime/src/sysvar_cache.rs]: Yeah, I was going to comment on the vectors vs. arrays in this PR as well, but you're right I think it may have to be vectors for that reason.

> i guess we can also cover that case by making size bigger than it needs to be.

Yeah I think we definitely should have this in the tests somewhere, since FD was pretty big on this point in the discussions.

File: program-runtime/src/sysvar_cache.rs
Comment (buffalojoec) [reply program-runtime/src/sysvar_cache.rs]: Ohh I see what you mean. Yeah, maybe what you have is fine then!

**Comment by buffalojoec** [comment]
> i dont believe anything outside of this repo should have any reason to call these functions, but i wonder if its unwise to make that assumption

To me this is a safe assumption, but definitely want to get a second opinion from the runtime team.

File: programs/bpf_loader/src/syscalls/sysvar.rs
Comment (2501babe) [reply programs/bpf_loader/src/syscalls/sysvar.rs:197]: to be clear, when the spec says that it "returns" 1, does that mean it should return `Ok(1)` rather than `Err(InstructionError::Custom(1))`? i assume the former because you said "graceful" but this isnt perfectly clear to me from the spec. does this mean passing in 0 for the length should return `Ok(0)` and do nothing? what do we return if the output buffer is too small? what about overflow of adding length to offset, even though that shouldnt be possible to hit? can we return errors at all?

Comment (2501babe) [reply programs/bpf_loader/src/syscalls/sysvar.rs]: to be clear, when the spec says that it "returns" 1, does that mean it should return `Ok(1)` rather than `Err(InstructionError::Custom(1))`? i assume the former because you said "graceful." does this mean passing in 0 for the length should return `Ok(0)` and do nothing?

Comment (2501babe) [reply programs/bpf_loader/src/syscalls/sysvar.rs:197]: wait a second, returning an error doesnt "abort the virtual machine" either does it...? i actually cant tell what the `ThisError` conversion turns them into either... `SyscallError` maybe? does this mean by the spec we should `panic!` on the failure cases...? and then is returning `Err` a "graceful success"......? im very confused by this and dont understand why we dont return errors uniformly including for the "successful failure" cases

Comment (2501babe) [reply programs/bpf_loader/src/syscalls/sysvar.rs:197]: ok well as best i can tell after making a test program and putting a `?` in the syscall for `Clock::get` the error is transformed into `SyscallError` the the declare macro and aborts computation somewhere i havent been able to locate through the indirection. so its not quite as complicated as i thought. im still skeptical of returning success codes for the soft failures tho because of implementation risks, particularly if safety mechanisms in rust itself result in a hard failure for agave for what would be a soft failure on firedancer

Comment (2501babe) [reply programs/bpf_loader/src/syscalls/sysvar.rs:197]: lol actually i think its fine, ive been thinking about this forwards (check the hard errors first because theyre more important) when i should have been thinking about this backwards (check the soft errors first because they cant happen accidentally due to rust). the specific thing i was afraid of was "check for possible hard error -> check for possible soft error -> try to do something that causes a hard error despite checks"

Comment (2501babe) [reply programs/bpf_loader/src/syscalls/sysvar.rs]: ive been wrestling with the consensus implications of having soft and hard errors for a few hours, wrote a test program to check actual behavior, wrote a few comments here, deleted them for being confused and tentative, and left a more actionable comment on the simd

honestly i dislike the multiple kinds of error tho and would be happier if we could just abort on the cases that the spec wants to return codes

Comment (buffalojoec) [reply programs/bpf_loader/src/syscalls/sysvar.rs]: Those are great points. I think we need to hash it out in the SIMD PR. I'll comment over there and we can circle back here!

**Comment by 2501babe** [comment]
@buffalojoec i pushed changes related to the error conditions in simd0127, without rebasing yet so the diff you see from last review is clean. the main change otherwise is i got rid of `read_sysvar_into` because the reading logic is now heavily dependent on error logic defined by the spec

tests will fail because of dependency vulns. once you approve ill rebase and ask for a second reviewer since we have been both so invested in this for the past month that i think outside eyes are needed. but i think we are almost at the finish line on this one!

File: programs/bpf_loader/src/syscalls/sysvar.rs
Comment (buffalojoec) [reply programs/bpf_loader/src/syscalls/sysvar.rs]: Circling back from SIMD conversations, the behavior here should be to abort the VM if any VM-abort constraints are violated, even if a graceful error constraint is also violated at the same time. Only if all VM-abort constraints are respected should we attempt graceful errors.

https://github.com/solana-foundation/solana-improvement-documents/pull/127#discussion_r1571071566

Comment (2501babe) [reply programs/bpf_loader/src/syscalls/sysvar.rs]: just checking, is this comment just to clarify resolution for future readers? the code does this now, you might not be seeing the latest commit if it looks otherwise (eg if you clicked the link to the file above this comment thread)

Comment (buffalojoec) [reply programs/bpf_loader/src/syscalls/sysvar.rs]: Yes, I'll mark resolved once I take a look, but yeah just for reader to see the conclusion.

File: program-runtime/src/sysvar_cache.rs
Comment (buffalojoec) [inline program-runtime/src/sysvar_cache.rs]: y??

File: programs/bpf_loader/src/syscalls/sysvar.rs
Comment (buffalojoec) [inline programs/bpf_loader/src/syscalls/sysvar.rs]: Solana nodes have to run a 64-bit machine, right? 😅

Referencing specifically `as usize`.

File: programs/bpf_loader/src/syscalls/sysvar.rs
Comment (buffalojoec) [inline programs/bpf_loader/src/syscalls/sysvar.rs]: This implementation looks great!

File: programs/bpf_loader/src/syscalls/mod.rs
Comment (buffalojoec) [inline programs/bpf_loader/src/syscalls/mod.rs]: You still need to clone the deserialized object in order to deal with the non-zero padding? What happens if you don't clone the resulting `EpochSchedule` here? Isn't the cloning happening _after_ all the bytes are read and deserialization finishes?

File: programs/bpf_loader/src/syscalls/mod.rs
Comment (buffalojoec) [inline programs/bpf_loader/src/syscalls/mod.rs]: Can `bincode::serialize_into` be used here?

File: programs/bpf_loader/src/syscalls/mod.rs
Comment (buffalojoec) [inline programs/bpf_loader/src/syscalls/mod.rs]: Nice test!

**Review by buffalojoec — COMMENTED** [review]
This lgtm, I just left a few tiny comments/questions. Nice work!!

File: programs/bpf_loader/src/syscalls/sysvar.rs
Comment (2501babe) [reply programs/bpf_loader/src/syscalls/sysvar.rs]: if not we would all be dead i think
```
hana@laptop:solana$ find . -name '*.rs' | xargs grep -n 'as usize' | wc
   1240    8031  147459
```

File: program-runtime/src/sysvar_cache.rs
Comment (2501babe) [reply program-runtime/src/sysvar_cache.rs]: oh im just removing mutability for the sake of the test so we can be sure nothing further can change it after this

Comment (2501babe) [reply program-runtime/src/sysvar_cache.rs]: (more precisely the creation of an immutable reference means the compiler enforces that the mutable reference doesnt live past this point regardless of the name shadowing)

File: programs/bpf_loader/src/syscalls/mod.rs
Comment (2501babe) [reply programs/bpf_loader/src/syscalls/mod.rs]: the test is doing the same thing as this for the same reason (the test doesnt go through the sysvar-specific syscalls, its purpose is to ensure getting those sysvars through the new syscall functions identically)
```
    // this clone looks unecessary, but it exists to zero out trailing alignment bytes
    // it is unclear whether this should ever matter
    // but there are tests using MemoryMapping that expect to see this
    // we preserve the previous behavior out of an abundance of caution
    *var = sysvar?.clone();
```

if we dont clone, the object works perfectly normal. the object equality assert would pass
```
assert_eq!(epochschedule_from_buf, src_epochschedule);
```

but byte equality would fail (`are_bytes_equal` uses raw pointers to step through the full buffer)
```
            assert!(are_bytes_equal(
                &epochschedule_from_buf,
                &clean_epochschedule
            ));
```

as noted in the comment, i dont think this should ever matter. in fact because the runtime doesnt care about the contents of the out buffer and just wants the object, it should never matter. im just being (perhaps excessively) paranoid about the fact we cant feature gate this so it has to behave identically

File: programs/bpf_loader/src/syscalls/mod.rs
Comment (2501babe) [reply programs/bpf_loader/src/syscalls/mod.rs]: oh yes youre right

File: programs/bpf_loader/src/syscalls/mod.rs
Comment (buffalojoec) [reply programs/bpf_loader/src/syscalls/mod.rs]: I think it's totally reasonable to go through those steps, no problem. Thanks for explaining!

**Review by buffalojoec — DISMISSED** [review]
Nice!! Looks great!

@Lichtso can you check out the changes to the Sysvar Cache here?
https://github.com/anza-xyz/agave/pull/560/files#diff-1aad504fccaea61fe9c2992f5030961c0734f65ea167b825d2ac505f8a4b986dR29-R42

File: program-runtime/src/sysvar_cache.rs
Comment (Lichtso) [inline program-runtime/src/sysvar_cache.rs]: Should be either a `match` or use the `::check_id()`

File: program-runtime/src/sysvar_cache.rs
Comment (Lichtso) [reply program-runtime/src/sysvar_cache.rs]: Also, all of these can be inlined on the callsite now, but can happen in a separate PR too.

File: programs/bpf_loader/src/syscalls/sysvar.rs
Comment (Lichtso) [reply programs/bpf_loader/src/syscalls/sysvar.rs]: Uff, does the SIMD specify the expected behavior?

File: program-runtime/src/sysvar_cache.rs
Comment (Lichtso) [inline program-runtime/src/sysvar_cache.rs:28]: Do we have a measurement of the actual size allocated across the entire struct here? Because now it is no longer shared between banks but cloned, right?

Comment (2501babe) [reply program-runtime/src/sysvar_cache.rs:28]: i dug into this last month when discussing `SysvarCache` with firedancer while prepping the simd and as best i can tell the object is created fresh for every bank

`_new_from_parent` sets `transaction_processor: TransactionBatchProcessor::default(),` when it instantiates a new bank, and the processer contains the `SysvarCache`. later, `_new_from_parent` updates the account data for the sysvars, then near the end calls `transaction_processor.fill_missing_sysvar_cache_entries` which routes to `sysvar_cache.fill_missing_entries`, which loads the account data using a callback via `banks_client` and assigns it to `SysvarCache`

the other reason to be quite sure `SysvarCache` is never shared or reused is because the setter functions only do the assignment if the entry is `None`. in other words if the cache _were_ reused in any way, the clock wouldnt tick

**Comment by 2501babe** [comment]
@Lichtso updated to use `check_id`

**Comment by Lichtso** [comment]
Why are there any changes to the existing sysvar syscall tests? These should stay stable

**Comment by 2501babe** [comment]
assuming you mean things like
```
        assert_eq!(
            cached_clock.expect("clock sysvar missing in cache"),
            clock.into()
        );
        assert_eq!(cached_clock.expect("clock sysvar missing in cache"), clock);
```

rather than the additions to cover new cases, this is because we changed the getters to no longer return an `Arc` because its no longer needed. we had been debating this earlier, if you think it would be better to maintain the current interface we could wrap our returns in `Arc`s? happy to make that change

**Comment by Lichtso** [comment]
For the sake of reviewing it I think it makes sense to split it in two PRs:
One for the internal interface changes, and one for the external interface changes.

**Comment by 2501babe** [comment]
ok, ive updated to preserve the previous `Arc` interface. the remaining changes to sysvar tests are just to add to them, so we can ensure the new syscall will function the same as the old ones

**Comment by Lichtso** [comment]
What I meant was to separate the addition of `declare_builtin_function!(SyscallGetSysvar,` (the change to the external interface) into its own PR.

**Comment by 2501babe** [comment]
ok no problem, ive updated this pr to have only the changes to `SysvarCache`, and opened a new pr #1307 containing the rest of the changes. i kept the `Arc` change so that none of the callers need to change

**Review by Lichtso — APPROVED** [review]

**Pull request merged**
